internal/manifest/manifest.go

```diff
@@ -0,0 +1,245 @@
+package manifest
+
+import (
+	"fmt"
+	"os"
+	"slices"
+
+	"github.com/klauspost/compress/zstd"
+
+	"github.com/canonical/chisel/internal/jsonwall"
+	"github.com/canonical/chisel/internal/setup"
+)
+
+type Package struct {
+	Kind    string `json:"kind"`
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Digest  string `json:"sha256"`
+	Arch    string `json:"arch"`
+}
+
+type Slice struct {
+	Kind string `json:"kind"`
+	Name string `json:"name"`
+}
+
+type Path struct {
+	Kind      string   `json:"kind"`
+	Path      string   `json:"path"`
+	Mode      string   `json:"mode"`
+	Slices    []string `json:"slices"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice"`
+	Path  string `json:"path"`
+}
+
+type Manifest struct {
+	db *jsonwall.DB
+}
+
+func Read(absPath string) (manifest *Manifest, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	file, err := os.OpenFile(absPath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	r, err := zstd.NewReader(file)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	jsonwallDB, err := jsonwall.ReadDB(r)
+	if err != nil {
+		return nil, err
+	}
+
+	manifest = &Manifest{db: jsonwallDB}
+	err = validate(manifest)
+	if err != nil {
+		return nil, err
+	}
+	return manifest, nil
+}
+
+func (manifest *Manifest) IteratePath(prefix string, f func(Path) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	iter, err := manifest.db.IteratePrefix(map[string]string{"kind": "path", "path": prefix})
+	if err != nil {
+		return err
+	}
+	for iter.Next() {
+		var path Path
+		err := iter.Get(&path)
+		if err != nil {
+			return err
+		}
+		err = f(path)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (manifest *Manifest) IteratePkgs(f func(Package) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	iter, err := manifest.db.Iterate(map[string]string{"kind": "package"})
```

Comment:
Although we don't have the ordering issue here, let's use the type as well for consistency.